### ISSUE TITLE
No writing sheet on ARMv7 and 32bit devices.

### ISSUE THREAD

**Issue by ThierryM** [xournalpp/xournalpp#2514]
**Affects versions :**
 - OS: Raspberry Pi OS
 - version of libgtk 3.24.5
 - Version of Xournal++: 1.1.0+dev
 - Installation method: builded directly according the Linux Build for Ubuntu/Debian

**Describe the bug**
When you open Xournal++, the zoom is fixed at 274% : no page, the background is grey so you think that the program doesn't work. You have to decrease at 100% to see the page.

**To Reproduce**
1. Launch Xournal++
2. The background of work surface is grey (no white page as expected)
3. Decrease the zoom level at 100%
3. You can see now the white page.

**Expected behavior**
Display the page all the time.

**Comment by Febbe** [comment]
I can only reproduce this, if the bookmode is enabled. 
But the issue is, that the zoom to fit option does not take the 2nd page into account.

**Comment by ThierryM** [comment]
If i have:
- 2 pages, nothing displayed if zoom level > 49%,
- 3 pages, nothing displayed if zoom level > 32%,
- Since 4 pages, nothing is displayed (zoom level limited at 30%).

**Comment by Febbe** [comment]
Can you leave us your xournalpp config files and a test document, or does this also happen with new documents?
Since you are using Raspberry Pi OS, we may not be able to reproduce or find the issue/bug. Then you have to debug the issue to fix it :/ .

**Comment by ThierryM** [comment]
It happens with new documents. I've just launched Xournal++ after I builded for testing.
It seems that the zoom level depends on the windows size of Xournal++ : 274% is when the windows is maximum. If I reduce the windows the zoom level decrease too (see the joined files).
![2020-12-09_18-18](https://user-images.githubusercontent.com/1470067/101684038-13e85300-3a66-11eb-8829-d05036edd831.png)
![2020-12-09_18-17](https://user-images.githubusercontent.com/1470067/101684047-15198000-3a66-11eb-8a01-82de9aff9ebd.png)

I join the settings.xml file too (renamed in settings.txt for uploading).
[settings.txt](https://github.com/xournalpp/xournalpp/files/5668520/settings.txt)

**Comment by Febbe** [comment]
Unfortunately there is everything fine with your settings file. Can you test another compiler (clang <-> gcc).

**Comment by emaninspace** [comment]
Version 1.0.10 does not seem to suffer from this issue. Version 1.0.17 however does fail to zoom past 100%.

**Comment by Febbe** [comment]
> Version 1.0.10 does not seem to suffer from this issue. Version 1.0.17 however does fail to zoom past 100%.

Do you know how to git bisect?
It would be a great help if you can find the first commit this occurs.
We can't reproduce this, because we don't have an ARM-64bit device.

I think it is a casting issue caused by undefined behavior.
You could also try to compile xournalpp with `-O1 -g -fsanitize=undefined -fsanitize=address -fno-omit-frame-pointer`

Other options can be found on:
https://clang.llvm.org/docs/UndefinedBehaviorSanitizer.html

Make sure to install the newest clang version and the newest ASAN lib.

**Comment by Febbe** [comment]
Can you print `g++ --version` and  `clang --version`

**Comment by vaghettiroberto** [comment]
vaghetti@desktop-raspberrypi:~/xournalpp/build/src $ g++ --version 
g++ (Raspbian 8.3.0-6+rpi1) 8.3.0
Copyright (C) 2018 Free Software Foundation, Inc.
This is free software; see the source for copying conditions.  There is NO
warranty; not even for MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.

**Comment by emaninspace** [comment]
I performed the bisect. Steps followed:

1) Cloned the repository. 
2) I knew that 1.0.12 did not display the unexpected behavior so I found the commit on the releases page which was cc44b20. 
3) Ran `git bisect start HEAD cc44b20`. 
4) Ran the build process and verified that the commit was not displaying the unexpected behavior.
5) Ran the bisect process building and checking.

6317b2a2b0a4c0737f3a5d659a295353753453cd is the first bad commit
commit 6317b2a2b0a4c0737f3a5d659a295353753453cd
Author: Fabian Keßler <fabian_kessler@gmx.de>
Date:   Wed Aug 7 19:58:06 2019 +0200

    changed some 32 bit int types to corresponding 64bit type

:040000 040000 b4fb90e50532a8a4d7bed727259b1333a4232570 328376562a25a4f3d3ecab897df3a49ff6cad57c M	src

**Comment by Febbe** [comment]
> 
> 
> EDIT: I didn't read carefully about the 64bit dependency. I installed 64 bit pi os and it compiles and runs just fine. 1.0.12 up to the commit below compiles fine on 32 bit pi os.
> 
> I performed the bisect. Steps followed:
> 
>     1. Cloned the repository.
> 
>     2. I knew that 1.0.12 did not display the unexpected behavior so I found the commit on the releases page which was [cc44b20](https://github.com/xournalpp/xournalpp/commit/cc44b206b75eac2cce98b3124d7e6292c5492614).
> 
>     3. Ran `git bisect start HEAD cc44b20`.
> 
>     4. Ran the build process and verified that the commit was not displaying the unexpected behavior.
> 
>     5. Ran the bisect process building and checking.
> 
> 
> [6317b2a](https://github.com/xournalpp/xournalpp/commit/6317b2a2b0a4c0737f3a5d659a295353753453cd) is the first bad commit
> commit [6317b2a](https://github.com/xournalpp/xournalpp/commit/6317b2a2b0a4c0737f3a5d659a295353753453cd)
> Author: Fabian Keßler [fabian_kessler@gmx.de](mailto:fabian_kessler@gmx.de)
> Date: Wed Aug 7 19:58:06 2019 +0200
> 
> ```
> changed some 32 bit int types to corresponding 64bit type
> ```
> 
> :040000 040000 b4fb90e50532a8a4d7bed727259b1333a4232570 328376562a25a4f3d3ecab897df3a49ff6cad57c M src

Seems like gcc is either producing 32-bit ARM Code, or it defines some datatypes like int/long differently. This results in incompatible and undefined casting behavior.
Can you try to compile xournalpp with sanitizers and/or `-Wconversion`

**Comment by Febbe** [comment]
Can you test if #2576 fixes the issue?

**Comment by emaninspace** [comment]
How do I add in the compile flag? If you tell me how to add the compile flag to the cmake instruction then I can test it.

**Comment by Febbe** [comment]
ok I made some changes to the code. They may fix it. 
To test my changes do the following
1. Go to the project directory 
2. open the terminal there.
3. add my repo with `git remote add Febbe https://github.com/Febbe/xournalpp.git`
4. fetch all changes: `git fetch --all`
5. checkout my changes `git checkout try_fix_no_page`
6. To build, follow the normal instructions how to build.

To test it with the compilation flag modify the cmake command like that:
`cmake <CMakeLists.txt's dir> -DCMAKE_CXX_FLAGS=-WConversion` 
You could also export this flag to your CXX_FLAGS variable:
`export CXX_FLAGS=-WConversion`

Search for all warnings regarding conversions in files like Layout ZoomControll or ScrollHandler.
Thank you for your help making Xournal++ a better software :).

**Comment by emaninspace** [comment]
It works.

**Comment by ThierryM** [comment]
Hi,
It seems that the bug is fixed : the page appears whatever the zoom rate on Raspberry Pi OS on a Raspberry Pi 4.
Thanks for the work !
Regards

**Comment by Febbe** [comment]
OK good to know. It seems like, size_t is not defined as 64bit unsigned integral type. Even If the system is a 64bit machine. 

I will refactor the fix and it must be reviewed. I try to merge it before Silvester.

**Comment by emaninspace** [comment]
Are the changes merged into the mainline (not sure about appropriate vocab)?

**Comment by Febbe** [comment]
The fix currently is blocked by 2 other fixes. Since I have good experience resolving merge conflicts, I will merge this last.

**Comment by Febbe** [comment]
@emaninspace @ThierryM can you retest the newest changes?
